test/azure/terraform_azure_vm_example_test.go

```diff
@@ -0,0 +1,191 @@
+
+func TestTerraformAzureVmExample(t *testing.T) {
+	t.Parallel()
+
+	// subscriptionID is overridden by the environment variable "ARM_SUBSCRIPTION_ID"
+	subscriptionID := ""
+	uniquePostfix := random.UniqueId() // "resource" //
+	expectedVmAdminUser := "testadmin"
+	expectedVMSize := compute.VirtualMachineSizeTypes("Standard_B1s")
+	expectedImageSKU := "2019-Datacenter-Core-smalldisk"
+	expectedImageVersion := "latest"
+	expectedDiskType := "Standard_LRS"
+	expectedSubnetAddressRange := "10.0.17.0/24"
+	expectedPrivateIPAddress := "10.0.17.4"
+	expectedManagedDiskCount := 1
+	expectedNicCount := 1
+
+	// Configure Terraform setting up a path to Terraform code.
+	terraformOptions := &terraform.Options{
+		// The path to where our Terraform code is located.
+		TerraformDir: "../../examples/azure/terraform-azure-vm-example",
+
+		// Variables to pass to our Terraform code using -var options.
+		Vars: map[string]interface{}{
+			"postfix":          uniquePostfix,
+			"user_name":        expectedVmAdminUser,
+			"vm_size":          string(expectedVMSize),
+			"vm_image_sku":     expectedImageSKU,
+			"vm_image_version": expectedImageVersion,
+			"disk_type":        expectedDiskType,
+			"private_ip":       expectedPrivateIPAddress,
+			"subnet_prefix":    expectedSubnetAddressRange,
+		},
+	}
+
+	// At the end of the test, run `terraform destroy` to clean up any resources that were created.
+	defer terraform.Destroy(t, terraformOptions)
+
+	// Run `terraform init` and `terraform apply`. Fail the test if there are any errors.
+	terraform.InitAndApply(t, terraformOptions)
+
+	// Run `terraform output` to get the values of output variables.
+	resourceGroupName := terraform.Output(t, terraformOptions, "resource_group_name")
+	virtualMachineName := terraform.Output(t, terraformOptions, "vm_name")
+	expectedVNetName := terraform.Output(t, terraformOptions, "virtual_network_name")
+	expectedSubnetName := terraform.Output(t, terraformOptions, "subnet_name")
+	expectedPublicAddressName := terraform.Output(t, terraformOptions, "public_ip_name")
+	expectedNicName := terraform.Output(t, terraformOptions, "network_interface_name")
+	expectedAvsName := terraform.Output(t, terraformOptions, "availability_set_name")
+	expectedOSDiskName := terraform.Output(t, terraformOptions, "os_disk_name")
+	expectedDiskName := terraform.Output(t, terraformOptions, "managed_disk_name")
+
+	// Comment for PR, to be removed:
+	// Please let me know if there are too many tests or alternate examples, happy to reduce the
+	// complexity and amount of code to be maintained. I tried to illustrate different approaches
+	// we have used in various scenarios to illustrate the flexability of Terratest.
+
+	t.Run("Strategies", func(t *testing.T) {
+		// Check the VM Size directly.
+		actualVMSize := azure.GetVirtualMachineSize(t, virtualMachineName, resourceGroupName, subscriptionID)
+		assert.Equal(t, expectedVMSize, actualVMSize)
+
+		// Check the VM Size by reference alternate example.
```

Comment:
NIT: What do you mean by "alternate example"?

Reply:
I labeled it as alternate since the three asserts in this section "Strategies" test for the same VM property "VM Size" using different patterns:
- getPropertyOfVM
- getVM then vm.Property
-getVMInstance then vmInstance.GetProperty()

I added the 2 alternates to show the how properties can be accessed on this complex resource while reducing the underlying calls to the sdk by reusing the same VM reference or instance. I did not create any other tests to support the alternate examples since these are meant to give the audience flexibility if they need to create their own tests.

If that does not align with the module and causes overhead I can remove these and stick to one pattern, getPropertyOfVM.

Reply:
Ah gotcha! The examples are fine, but I was confused by the comment. I suggest rewording to:

> An alternate example for checking the VM size by reference

As written, I grouped "reference alternate example" together which made it confusing, instead of reading it as "Check the VM Size by reference - alternate example".

Reply:
Fixed.

Reply:
Updated along with the sub test breakout.